File: examples/test_libmeshb_block.c

/* libMeshb 7.2 basic example: read a quad mesh, split it into triangles
    and write the result back using fast block transfer */

#include <stdio.h>
#include <stdlib.h>
#include <libmeshb7.h>


int main()
{
    int i, NmbVer, NmbQad, ver, dim, *RefTab, (*QadTab)[5], (*TriTab)[4];
    long long InpMsh, OutMsh;
    float (*VerTab)[3];


    /*-----------------------------------*/
	/* Open mesh file "quad.meshb"       */
    /*-----------------------------------*/

    if(!(InpMsh = GmfOpenMesh("quad.meshb", GmfRead, &ver, &dim)))
        return(1);

    printf("InpMsh : idx = %lld, version = %d, dimension = %d\n", InpMsh, ver, dim);

    if(dim != 3)
        exit(1);

	/* Read the number of vertices and allocate memory */
    NmbVer = GmfStatKwd(InpMsh, GmfVertices);
    printf("InpMsh : nmb vertices = %d\n", NmbVer);
    VerTab = malloc((NmbVer+1) * 3 * sizeof(float));
    RefTab = malloc((NmbVer+1) * sizeof(int));

	/* Read the number of quads and allocate memory */
    NmbQad = GmfStatKwd(InpMsh, GmfQuadrilaterals);
    printf("InpMsh : nmb quads = %d\n", NmbQad);
    QadTab = malloc((NmbQad+1) * 5 * sizeof(int));
    TriTab = malloc((NmbQad+1) * 2 * 4 * sizeof(int));

	/* Read the vertices */
    GmfGetBlock(InpMsh, GmfVertices, 1, NmbVer, NULL, \
                GmfFloat, &VerTab[1][0], &VerTab[ NmbVer ][0], \
                GmfFloat, &VerTab[1][1], &VerTab[ NmbVer ][1], \
                GmfFloat, &VerTab[1][2], &VerTab[ NmbVer ][2], \
                GmfInt,   &RefTab[1],    &RefTab[ NmbVer ] );

	/* Read the quads */
    GmfGetBlock(InpMsh, GmfQuadrilaterals, 1, NmbQad, NULL, \
                GmfInt, &QadTab[1][0], &QadTab[ NmbQad ][0], \
                GmfInt, &QadTab[1][1], &QadTab[ NmbQad ][1], \
                GmfInt, &QadTab[1][2], &QadTab[ NmbQad ][2], \
                GmfInt, &QadTab[1][3], &QadTab[ NmbQad ][3], \
                GmfInt, &QadTab[1][4], &QadTab[ NmbQad ][4] );

	/* Close the quad mesh */
    GmfCloseMesh(InpMsh);


    /*-----------------------------------*/
    /* Create the triangluated mesh      */
    /*-----------------------------------*/

    for(i=1;i<=NmbQad;i++)
    {
        TriTab[i*2-1][0] = QadTab[i][0];
        TriTab[i*2-1][1] = QadTab[i][1];
        TriTab[i*2-1][2] = QadTab[i][2];
        TriTab[i*2-1][3] = QadTab[i][4];

        TriTab[i*2][0] = QadTab[i][0];
        TriTab[i*2][1] = QadTab[i][2];
        TriTab[i*2][2] = QadTab[i][3];
        TriTab[i*2][3] = QadTab[i][4];
    }


	/*-----------------------------------*/
	/* Write the triangle mesh           */
	/*-----------------------------------*/

    if(!(OutMsh = GmfOpenMesh("tri.meshb", GmfWrite, ver, dim)))
        return(1);

	/* Write the vertices */
    GmfSetKwd(OutMsh, GmfVertices, NmbVer);
    GmfSetBlock(OutMsh, GmfVertices, NULL, \
                GmfFloat, &VerTab[1][0], &VerTab[ NmbVer ][0], \
                GmfFloat, &VerTab[1][1], &VerTab[ NmbVer ][1], \
                GmfFloat, &VerTab[1][2], &VerTab[ NmbVer ][2], \
                GmfInt,   &RefTab[1],    &RefTab[ NmbVer ] );

	/* Write the triangles */
    GmfSetKwd(OutMsh, GmfTriangles, 2*NmbQad);
    GmfSetBlock(OutMsh, GmfTriangles, NULL, \
                GmfInt, &TriTab[1][0], &TriTab[ 2*NmbQad ][0], \
                GmfInt, &TriTab[1][1], &TriTab[ 2*NmbQad ][1], \
                GmfInt, &TriTab[1][2], &TriTab[ 2*NmbQad ][2], \
                GmfInt, &TriTab[1][3], &TriTab[ 2*NmbQad ][3] );

	/* Do not forget to close the mesh file */
    GmfCloseMesh(OutMsh);
    printf("OutMsh : nmb triangles = %d\n", 2*NmbQad);

    free(QadTab);
    free(TriTab);
    free(RefTab);
    free(VerTab);

    return(0);
}

File: examples/test_libmeshb_block.f

c     libMeshb 7.79 example: transform a quadrilateral mesh into a triangular one
c     using fast block transfer

      include 'libmeshb7.ins'

      integer n
      parameter (n=4000)
      integer i, ver, dim, res, NmbVer, NmbQad
     +, RefTab(n), TriTab(4,2*n), QadTab(5,n)
      integer t(1),d,ho,s
      integer*8 InpMsh, OutMsh, m(1)
      real*8 VerTab(3,n)


c     --------------------------------------------
c     Open the quadrilateral mesh file for reading
c     --------------------------------------------

      InpMsh = gmfopenmeshf77('../sample_meshes/quad.meshb'
     +,GmfRead,ver,dim)

      if(InpMsh.eq.0) STOP ' InpMsh = 0'
      if(dim.ne.3) STOP ' dimension <> 3'

c     Check memory bounds
      NmbVer = gmfstatkwdf77(InpMsh, GmfVertices, 0, s, t, 0, ho)
      if(NmbVer.gt.n) STOP 'Too many vertices'

      NmbQad = gmfstatkwdf77(InpMsh, GmfQuadrilaterals, 0, s, t, 0, ho)
      if(NmbQad.gt.n) STOP 'Too many quads'

c     Print some information on the open file
      print*, 'input mesh  :', InpMsh
      print*, 'version     :', ver
      print*, 'dimension   :', dim
      print*, 'vertices    :', NmbVer
      print*, 'quads       :', NmbQad

c     Read the vertices using a vector of 3 consecutive doubles
c     to store the coordinates
      res = gmfgetvertices(InpMsh,
     +        1, NmbVer, 0, m,
     +        VerTab(1,1), VerTab(1,NmbVer),
     +        RefTab(  1), RefTab(  NmbVer))

c     Read the quads using one single vector of 5 consecutive integers
      res = gmfgetelements(InpMsh, GmfQuadrilaterals,
     +        1, NmbQad, 0, m,
     +        QadTab(1,1), QadTab(1,NmbQad),
     +        QadTab(5,1), QadTab(5,NmbQad))

c     Close the quadrilateral mesh
      res = gmfclosemeshf77(InpMsh)


c     -------------------------------------------
c     Convert the quad mesh into a triangular one
c     -------------------------------------------

      do i = 1,2*NmbQad
         if(mod(i,2) .EQ. 1) then
            TriTab(1,i) = QadTab(1,(i+1)/2)
            TriTab(2,i) = QadTab(2,(i+1)/2)
            TriTab(3,i) = QadTab(3,(i+1)/2)
            TriTab(4,i) = QadTab(5,(i+1)/2)
         else
            TriTab(1,i) = QadTab(1,(i+1)/2)
            TriTab(2,i) = QadTab(3,(i+1)/2)
            TriTab(3,i) = QadTab(4,(i+1)/2)
            TriTab(4,i) = QadTab(5,(i+1)/2)
         endif
      end do


c     -----------------------
c     Write a triangular mesh
c     -----------------------

      OutMsh = gmfopenmeshf77('tri.meshb', GmfWrite, ver, dim)
      if(OutMsh.eq.0) STOP ' OutMsh = 0'

c     Set the number of vertices
      res = gmfsetkwdf77(OutMsh, GmfVertices, NmbVer, 0, t, 0, ho)

c     Write them down using separate pointers for each scalar entry
      res = gmfsetvertices(OutMsh,
     +         1, NmbVer, 0, m,
     +         VerTab(1,1), VerTab(1,NmbVer),
     +         RefTab(1),   RefTab(NmbVer))

c     Write the triangles using 4 independant set of arguments
c     for each scalar entry: node1, node2, node3 and reference
      res = gmfsetkwdf77(OutMsh, GmfTriangles, 2*NmbQad, 0, t, 0, ho)
      res = gmfsetelements(OutMsh, GmfTriangles,
     +                  1, 2*NmbQad, 0, m,
     +                  TriTab(1,1), TriTab(1,2*NmbQad),
     +                  TriTab(4,1), TriTab(4,2*NmbQad))

c     Don't forget to close the file
      res = gmfclosemeshf77(OutMsh)

      print*, 'output mesh :',NmbVer,' vertices,',
     +         2*NmbQad,'triangles'

      end      
